src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DynamicFiles/BackgroundDocumentGenerator.cs

```diff
@@ -251,33 +253,27 @@ private void ReportError(IProjectSnapshot project, Exception ex)
 
     private bool Suppressed(IProjectSnapshot project, IDocumentSnapshot document)
     {
-        lock (_suppressedDocuments)
-        {
-            var filePath = document.FilePath.AssumeNotNull();
+        var filePath = document.FilePath.AssumeNotNull();
 
-            if (_projectManager.IsDocumentOpen(filePath))
-            {
-                _suppressedDocuments.Add(filePath);
-                _infoProvider.SuppressDocument(project.Key, filePath);
-                return true;
-            }
-
-            _suppressedDocuments.Remove(filePath);
-            return false;
+        if (_projectManager.IsDocumentOpen(filePath))
+        {
+            ImmutableInterlocked.Update(ref _suppressedDocuments, static (set, filePath) => set.Add(filePath), filePath);
```

Comment:
Is there a chance this could be called multiple times for the same file path? If so, you could test the return value of `Update` and skip the next step if false.